File: problems/profitabletrip/submissions/accepted/profitabletrip-zf.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

struct edge { int v, p; };

int main() {
  int n, m, w;
  cin >> n >> m >> w;

  vector<vector<edge>> g(n+1);
  while (m--) {
    int u, v, p;
    cin >> u >> v >> p;
    g[u].push_back({v,p});
  }

  vector<int> val(n+1, -1e9);

  queue<int> q;

  auto update = [&](int v, int x) {
    if (min(w,x) > val[v]) {
      val[v] = min(w,x);
      q.push(v);
    }
  };

  update(1, 0);

  while (!q.empty()) {
    int u = q.front();
    q.pop();

    for (auto [v,p] : g[u]) update(v, p+val[u]);
  }

  cout << val[n] << endl;
}

File: problems/profitabletrip/submissions/accepted/profitabletrip_zf.java
import java.util.*;

public class profitabletrip_zf {

    static class Edge {
        int v, p;

        Edge(int v, int p) {
            this.v = v;
            this.p = p;
        }
    }

    private int n, w;
    private int[] val;
    private Map<Integer, List<Edge>> g;
    private Queue<Integer> q;

    private void update(int v, int x) {
        if (Math.min(w, x) > val[v]) {
            val[v] = Math.min(w, x);
            q.add(v);
        }
    }

    private void work() {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        int m = sc.nextInt();
        w = sc.nextInt();

        g = new HashMap<>();

        while (m-- > 0) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int p = sc.nextInt();
            if (!g.containsKey(u)) g.put(u, new ArrayList<>());
            g.get(u).add(new Edge(v, p));
        }

        val = new int[n + 1];
        Arrays.fill(val, Integer.MIN_VALUE / 16);
        q = new LinkedList<>();

        update(1, 0);

        while (!q.isEmpty()) {
            int u = q.poll();
            if (g.containsKey(u)) {
                for (Edge e : g.get(u)) {
                    update(e.v, e.p + val[u]);
                }
            }
        }

        System.out.println(val[n]);
    }

    public static void main(String[] args) {
        new profitabletrip_zf().work();
    }
}
